lib/Doctrine/ORM/Internal/IdentityMap.php

```diff
@@ -0,0 +1,217 @@
+<?php
+
+namespace Doctrine\ORM\Internal;
+
+use Countable;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Doctrine\ORM\ORMInvalidArgumentException;
+
+use function array_map;
+use function array_sum;
+use function get_class;
+use function implode;
+use function in_array;
+
+class IdentityMap implements Countable
```

Comment:
I am thinking about implementing the `ArrayAccess` interface or extending from`SplObjectStorage`, but it's under discussion, the problem is that we need to use entity identifiers separately. If we can change the logic not to use entity identifieres, that could be possible. And since PHP8 we even can use `WeakMap`